Make .str/.dt available for Series of type category with string/datetime

jankatins opened the pull request:
If a series is a type category and the underlying Categorical has categories of type string or datetime, then make it possible to use the `.str`/`.dt` assessor on such a series.

The string/dt methods work on the categories (and therefore fast if we have
only a few categories), but return a Series with a dtype other than
category (integer, boolean, string,...), so that it is no different if we use
`.str` / `.dt` on a series of type string or of type category.

The main reason for that is that I think things like `s.str.slice(...) + s.str.slice(...)` should work.

Closes: https://github.com/pydata/pandas/issues/10661

jankatins committed:
Make .str available for Series of type category with strings

If a series is a type category and the underlying Categorical has
categories of type string, then make it possible to use the `.str`
assessor on such a series.

The string methods work on the categories (and therefor fast if we have
only a few categories), but return a Series with a dtype other than
category (boolean, string,...), so that it is no different if we use
`.str` on a series of type string or of type category.

JanSchulz commented:
I've no idea why `.cat` is not in `dir(series_of_type_category)`?!?

kawochen commented:
@JanSchulz 
https://github.com/pydata/pandas/blob/master/pandas/core/series.py#L2675

jankatins committed:
TST: Allow for more than one acessor on a Series

`.dt` and `.str` can be available when the Series is of type
category, which makes `.cat` available.

jankatins committed:
DOC: whatsnew and docs for multiple accessors

Also add some docs in text.rst to mention the performance gains
when using ``s_cat.str`` vs ``s.str``.

JanSchulz commented:
@kawochen Thanks! 7a20c36

jreback commented on pandas/core/base.py:
```diff
@@ -521,9 +521,14 @@ def searchsorted(self, key, side='left'):
-            # but that isn't practical for performance reasons until we have a
+        if isinstance(self, Series) and not(
+                    (com.is_categorical_dtype(self.dtype) and
+                     com.is_object_dtype(self.values.categories)) or
```
use ``self.cat.categories``, because otherwise you could be forcing a conversion here

JanSchulz replied:
Can do, but as I understand the code, I made sure that it is a categorical before using the categories on `values`: `com.is_categorical_dtype(self.dtype) and com.is_object_dtype(self.values.categories)`

jreback commented on pandas/core/strings.py:
```diff
@@ -1057,8 +1058,16 @@ class StringMethods(object):
-    def __init__(self, series):
-        self.series = series
+    def __init__(self, data):
+        if is_categorical_dtype(data):
```
```
self._is_categorical = is_categorical_dtype(dat)
self.data = data.cat.categories if self._is_categorial else data
```

JanSchulz replied:
done

jreback commented on pandas/tseries/common.py:
```diff
@@ -45,32 +45,47 @@ def maybe_to_datetimelike(data, copy=False):
 
     index = data.index
+    name = data.name
+    if is_categorical_dtype(data):
```
similar to how doing it in str above

JanSchulz replied:
Not sure if this is better, as it is two instead of one comparison:

```
 if is_categorical_dtype(data):
        orig = data
        data = orig.values.categories
    else:
        orig = None
```
vs
```
orig = data if  is_categorical_dtype(data) else None
if orig is not None:
        data = orig.values.categories
```

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -3625,6 +3625,39 @@ def test_cat_accessor_api(self):
+        s = Series(list('aabb')).astype('category')
+        self.assertIsInstance(s.str, StringMethods)
+
+        res = s.str.contains("a")
```
can you run thru all of the methods and assert their return types (e.g. boolean / object) (maybe better done in test_strings)

JanSchulz replied:
Isn't that already tested in the normal string methods and `take1d`?

JanSchulz replied:
ok, added one test for all return type I could find (bool, str, int, objects/lists).

jreback replied:
we need tests for all the methods as I suspect some don't work (or do they all?)

JanSchulz replied:
I can't think of a reason why they shouldn't: ’index.str' should be in the tests for '.str' and the only difference to the rest of the '.str' cases is the 'take1d'.

JanSchulz replied:
Maybe another testcase for expansion, e.g. 's.str.split()'

JanSchulz replied:
ok, that one found a bug (and a stupid guy in front of the keyboard who debugged 20min why `s.str.findall(" ") != c.str.split(" ")`...).

Anyway: under the condition that `take1d` and all the string methods are well tested, IMO this is now all tested as well: all new codepath are taken during the tests.

JanSchulz commented:
Yay, it's green :-)

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -3625,6 +3625,62 @@ def test_cat_accessor_api(self):
+        # https://github.com/pydata/pandas/issues/10661
+        from pandas.core.strings import StringMethods
+        s = Series(list('aabb'))
+        c = s.astype('category')
```
much better to have a loop here, that just looks at all of the StringMethods and tests against them, rather than writing out each one. (of course prob will need some special cases).

JanSchulz replied:
Im not so sure: you would need to replace this 6 3liners by a much harder to read list:

```
tests = [("method", (args), {"kwarg":"whatever"}),...]
```

I still don't think it should be necessary to build tests for all string methods, as far as I can tell, I covered all code paths (`_wrap_result` and `_wrap_result_expand`)

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -3625,6 +3625,62 @@ def test_cat_accessor_api(self):
+        from pandas.tseries.common import Properties
+        from pandas.tseries.index import date_range
+        s = Series(date_range('1/1/2015', periods=5))
+
```
same here, need to loop thru all of the methods/properties.

JanSchulz replied:
same here: IMO that thing is tested:

dt works on Index, so that path is covered and we basically do this here: take the categories, which are an index, do the dt things on it and the only difference to a dt on an index is, that it uses `take1d`on the result.

jreback replied:
I just want a test that it 'works' on all that ``.str`` works on (and corresponding where ``.dt`` works). I assume it will, but a test will ensure it.

jreback commented:
can you update the doc as well (maybe a reference / example in the Categorical section to one in the TextMethods section)

JanSchulz commented:
Added docs in categorical.rst

JanSchulz commented:
Ok, @jreback, you were right, it needed more tests...

Rebased + new tests + fixes... Lets wait what travis says...

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -3640,6 +3640,153 @@ def test_cat_accessor_no_new_attributes(self):
+        # Handcrafted by
+        # s = pandas.Series(list("abcd"))
+        # t = "('%s', (), {})"
+        # functs = [t % f for f in dir(s.str) if not f.startswith("_")]
```
so instead of listing these out as this would be instantly out-dated if something else is added.

create a list of the ones that need special arguments, then programatically create the rest

JanSchulz replied:
done

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -3640,6 +3640,153 @@ def test_cat_accessor_no_new_attributes(self):
+            else:
+                tm.assert_numpy_array_equal(res, exp)
+
+        attr_defs = ['date', 'day', 'dayofweek', 'dayofyear', 'days_in_month', 'daysinmonth',
```
all of these are defined in ``DatetimeIndex._datetimelike_ops`` (and ``Timesdelta/PeriodIndex``), so no need to actually list them (and defeats the purpose if things are added). Just iterate the already existing list, providing special cases if needed.

JanSchulz replied:
done

JanSchulz commented:
Ok, two failures: https://travis-ci.org/pydata/pandas/builds/91286698

I can't reproduce both errors (in py3.5 and py2.7, and it seems I'm too stupid to get a proper py3.4 environment which can run `python setup.py build_ext --inplace` :-( )

On py3.3/3.4:
`nosetests pandas.io.tests.test_data`
```
=====================================================================
ERROR: test_get_underlying_price (pandas.io.tests.test_data.TestYahooOptions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/util/testing.py", line 1774, in wrapper
    return t(*args, **kwargs)
  File "/home/travis/build/pydata/pandas/pandas/io/tests/test_data.py", line 400, in test_get_underlying_price
    quote_price = options_object._underlying_price_from_root(root)
  File "/home/travis/build/pydata/pandas/pandas/io/data.py", line 737, in _underlying_price_from_root
    underlying_price = root.xpath('.//*[@class="time_rtq_ticker Fz-30 Fw-b"]')[0]\
IndexError: list index out of range
```

Py27, https://travis-ci.org/pydata/pandas/jobs/91286713
`nosetests pandas.tseries.tests.test_timeseries`
```
======================================================================
FAIL: test_datetime64_with_DateOffset (pandas.tseries.tests.test_timeseries.TestDatetimeIndex)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/pydata/pandas/pandas/tseries/tests/test_timeseries.py", line 2668, in test_datetime64_with_DateOffset
    assert_func(klass([op + x for x in s]), op + s)
  File "/opt/python/2.7.9/lib/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "/home/travis/build/pydata/pandas/pandas/util/testing.py", line 2042, in assert_produces_warning
    % expected_warning.__name__)
AssertionError: Did not see expected warning of class 'PerformanceWarning'.
```

jreback commented:
ignore the first one, 2nd I am fixing now.

jankatins committed:
Make .dt available for Series of type category with datetimes

If a series is a type category and the underlying Categorical has
categories of type datetime, then make it possible to use the .dt
assessor on such a series.

The string methods work on the categories (and therefore fast if we have
only a few categories), but return a Series with a dtype other than
category (integer,...), so that it is no different if we use
.dt on a series of type datetime or of type category.

JanSchulz commented:
Ok, remodeled the tests and rebased. You could cancel the following travis run, the differnece to the latest is just some reshuffeling in the commits: https://travis-ci.org/pydata/pandas/builds/91400953

JanSchulz commented:
Yay, https://travis-ci.org/pydata/pandas/builds/91400953 is green so hopefull this one (no code change in the PR, only moved some changes to a different commit) will also be gree :-)

JanSchulz commented:
@jreback green... :-) Thanks for fixing whatever needed fixing :-)

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -3640,6 +3640,161 @@ def test_cat_accessor_no_new_attributes(self):
+            try:
+                res = getattr(c.str, func)(*args, **kwargs)
+                exp = getattr(s.str, func)(*args, **kwargs)
+            except Exception as e:
```
why are you try/catching then raise? (for debugging?)

JanSchulz replied:
Yes: if the methods failed due to missing args, then this was the easiest way to find out why they failed.

jreback commented on pandas/tests/test_categorical.py:
```diff
@@ -3640,6 +3640,161 @@ def test_cat_accessor_no_new_attributes(self):
+                    exp = getattr(s.dt, func)(*args, **kwargs)
+                except Exception as e:
+                    print(name, func)
+                    raise(e)
```
same here

JanSchulz replied:
The problem is even worse here if a arg is missing you can't find out form the log what's the problematic object and function, as the error is thrown be `getattr` in the test method and not by a method in one of the objects (where the stacktrace would give a hint)

JanSchulz commented on doc/source/categorical.rst:
```diff
@@ -515,6 +515,30 @@ To get a single value `Series` of type ``category`` pass in a list with a single
+
+.. versionadded:: 0.17.1
+
+The accessors  ``.dt`` and ``.str`` work if the ``s.cat.categories`` are of an appropriate type:
```
done

JanSchulz commented on doc/source/categorical.rst:
```diff
@@ -515,6 +515,30 @@ To get a single value `Series` of type ``category`` pass in a list with a single
+    work, it can be faster to convert the original ``Series`` to one of type ``category`` and use
+    ``.str.<method>`` or ``.dt.<property>`` on that.
+
+
```
I hope this is better:

```

    The work is done on the ``categories`` and then a new ``Series`` is constructed. This has
    some performance implication if you have a ``Series`` of type string, where lots of elements
    are repeated (i.e. the number of unique elements in the ``Series`` is a lot smaller than the 
    length of the ``Series``). In this case it can be faster to convert the original ``Series`` 
    to one of type ``category`` and use``.str.<method>`` or ``.dt.<property>`` on that.
```

JanSchulz commented on doc/source/categorical.rst:
```diff
@@ -515,6 +515,30 @@ To get a single value `Series` of type ``category`` pass in a list with a single
+
+.. note::
+
+    The work is done on the ``categories`` and then a new Series is constructed. If the number of
```
done

JanSchulz commented on doc/source/categorical.rst:
```diff
@@ -515,6 +515,30 @@ To get a single value `Series` of type ``category`` pass in a list with a single
+
 Setting
 ~~~~~~~
 
```
Added a new para to explain the return type.

JanSchulz commented:
@jreback Hopefully addressed your comments. Thanks for the review!

jorisvandenbossche commented on doc/source/categorical.rst:
```diff
@@ -515,6 +515,43 @@ To get a single value `Series` of type ``category`` pass in a list with a single
+    str_cat.str.contains("a")
+
+    date_cat = Series(date_range('1/1/2015', periods=5))
+    date_cat.dt.day
```
This will not yet be a categorical series

JanSchulz replied:
Whoops...

jorisvandenbossche commented on doc/source/categorical.rst:
```diff
@@ -515,6 +515,43 @@ To get a single value `Series` of type ``category`` pass in a list with a single
+
+.. ipython:: python
+
+    str_s = Series(list('aabb'))
```
can you use `pd.Series` here instead of `Series`?

JanSchulz replied:
done

jorisvandenbossche commented on doc/source/whatsnew/v0.17.1.txt:
```diff
@@ -64,6 +64,7 @@ Enhancements
   .. ipython:: python
 
      pd.Index([1, np.nan, 3]).fillna(2)
+- Series of type ``"category"`` now make ``.str.<...>`` and ``.dt.<...>`` accessor methods / properties available, if the categories are of that type. (:issue:`10661`)
```
A blank line is missing above this line (after the previous example)

JanSchulz replied:
done

jorisvandenbossche commented:
@JanSchulz Nice! I just added a few minor doc comments

JanSchulz commented:
@jorisvandenbossche Addressed your comments. Thanks for the review!

jreback commented on doc/source/categorical.rst:
```diff
@@ -515,6 +515,45 @@ To get a single value `Series` of type ``category`` pass in a list with a single
+.. note::
+
+    The work is done on the ``categories`` and then a new ``Series`` is constructed. This has
+    some performance implication if you have a ``Series`` of type string, where lots of elements
```
This is still not correct from a user perspective. What you want / need to know is that you are NOT getting a categorical type back.

The fact that this can equally be done ONLY on the categories and manually constructing a new categorical is important ,but should be separate from this.

jorisvandenbossche replied:
I think this is explained in the text above the example? 
But I agree it can maybe be repeated explicitly in the note, as this note is what stands out from the text

JanSchulz replied:
I wanted to have this note only on the performance implications, nothing to do with the return type. IMO this could also go into a note in the string method docs.

jreback replied:
yes, I think this would be gr8 on the doc-strings as well (you can do a follow up) for that if you'd like

JanSchulz replied:
Ok, will add the same para on to of text.rst

jreback commented on pandas/core/strings.py:
```diff
@@ -1080,6 +1083,15 @@ def __iter__(self):
+        # But for some opperations, we have to do the stuff on the full values,
+        # so make it possible to skip this step as the method already did this before
+        # the transformation...
+        use_codes = kwargs.get('use_codes', True)
```
why don't you make ``use_codes=False`` a keyword argument directly

JanSchulz replied:
I didn't because name wasn't a kwarg, too:

```
name = kwargs.get('name') or getattr(result, 'name', None) or self._orig.name
```

Or do I misunderstand something here?

jreback replied:
should prob just make ``name`` a kw too! these are all internal (I think), more clear to show the kw (if its needed)

JanSchulz replied:
Ok, will submit this in a new commit to see if the original PR still passes travis...

jreback commented on pandas/core/strings.py:
```diff
@@ -1123,36 +1140,41 @@ def _wrap_result_expand(self, result, expand=False):
+            return self._wrap_result(result)
+        else:
+            result = str_cat(self._orig, others=others, sep=sep, na_rep=na_rep)
+            return self._wrap_result(result, use_codes=False)
```
maybe

```
data = self._orig if self._is_categorical else self._data
return self._wrap_result(str_cat(data, .......), use_codes=self._is_categorical)
```

JanSchulz replied:
ok

jreback commented on pandas/core/strings.py:
```diff
@@ -1332,34 +1354,40 @@ def encode(self, encoding, errors="strict"):
-        result = str_get_dummies(self.series, sep)
-        return self._wrap_result(result)
+        if not self._is_categorical:
+            result = str_get_dummies(self._data, sep)
```
similar to ``str_cat`` as above

JanSchulz replied:
done

jreback commented:
couple comments. pls squash when finished.

JanSchulz commented on doc/source/categorical.rst:
```diff
@@ -515,6 +515,45 @@ To get a single value `Series` of type ``category`` pass in a list with a single
+The returned ``Series`` (or ``DataFrame``) is of the same type as if you used the
+``.str.<method>`` / ``.dt.<method>`` on a ``Series`` of that type. In the above case, the
+returned values of ``str_cat.str.contains("a")`` and ``str_s.str.contains("a")`` or
+``date_cat.dt.day`` and ``date_s.dt.day`` will be equal:
```
Would it be ok to make (half of ) this paragraph into note and insert a `(and not of type "category")` like this:

```
.. note::

    The returned ``Series`` (or ``DataFrame``) is of the same type as if you used the 
    ``.str.<method>`` / ``.dt.<method>`` on a ``Series`` of that type (and not of type ``category``). 

That means, that in the above case, the returned values of ``str_cat.str.contains("a")`` and ``str_s.str.contains("a")`` or ``date_cat.dt.day`` and ``date_s.dt.day`` will be equal:

...
```

JanSchulz commented:
@jreback: please no squash: I really put some work into making each commit stand on its own, the unit of change is each commit (4 of them...) and not the complete PR.

JanSchulz commented:
Ok, addressed the comments, please review the doc change if that is clearer now.

jreback commented:
ok, lgtm. ping on green.

on the squash; ok for this one :) normally we just squashem as a matter of course.

jreback commented:
@JanSchulz ok, ping when green (after any more doc changes you have)

JanSchulz commented:
Ok, added something to text.rst. Please review. It would alos good if someone with access to travis would cancel the other builds (https://travis-ci.org/pydata/pandas/builds/91604782 and https://travis-ci.org/pydata/pandas/builds/91605534)  so that we have 4 hours less to wait for this :-)

JanSchulz commented:
@jreback ping :-)

jreback commented:
merged via https://github.com/pydata/pandas/commit/6eefe756a9cdd8d8b9efc95920f8434245bba8cf

thanks!

as usual, pls review built docs for accuracy

JanSchulz commented:
@jreback Thanks for all the reviews and for your patience! The PR got much better thanks to you and @jorisvandenbossche!

Will add a PR forthis:

```
In [124]: date_s = pd.Series(date_range('1/1/2015', periods=5))
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-124-f7c79f210480> in <module>()
----> 1 date_s = pd.Series(date_range('1/1/2015', periods=5))

NameError: name 'date_range' is not defined
```

jreback commented:
@JanSchulz no thank you!

great PR as always from you.

I got the doc fix, was making some corrections anyhow.
